### ISSUE TITLE
cv::stereoRectify  gives different results between version 3.4.0 and 3.4.1

### ISSUE THREAD

**Issue by fbergama** [opencv/opencv#11131]
##### System information (version)

- OpenCV => 3.4.1  and 3.4.0
- Operating System / Platform => Linux Ubuntu artful 64 Bit
- Compiler => GCC 7.2.0

##### Detailed description

cv::stereoRectify in the latest 3.4.1 version of OpenCV is giving wrong results when alpha=1.0. In particular, many values are negated and the output ROIs are wrongly set to zero.

##### Steps to reproduce

 ```.cpp
#include <iostream>
#include <opencv2/core/core.hpp>
#include <opencv2/calib3d/calib3d.hpp>


int main( int argc, char* argv[] )
{
    cv::Mat K0 = (cv::Mat_<double>(3,3) << 1457.572438721727, 0, 1212.945694211622, 0, 1457.522226502963, 1007.32058848921, 0, 0, 1);
    cv::Mat kk0 =  cv::Mat_<double>::zeros(1,5);
    cv::Mat K1 = (cv::Mat_<double>(3,3) << 1460.868570835972, 0, 1215.024068023046, 0, 1460.791367088, 1011.107202932225, 0, 0, 1);
    cv::Mat kk1 =  cv::Mat_<double>::zeros(1,5);

    cv::Mat R = (cv::Mat_<double>(3,3) << 0.9985404059825475, 0.02963547172078553, -0.04515303352041626, -0.03103795276460111, 0.9990471552537432, -0.03068268351343364, 0.04420071389006859, 0.03203935697372317, 0.9985087763742083 );

    cv::Mat T = (cv::Mat_<double>(3,1) << 0.9995500167379527, 0.0116311595111068, 0.02764923448462666 );

    cv::Size imgsize( 2456, 2058 );

    cv::Mat R1;
    cv::Mat R2;
    cv::Mat P1;
    cv::Mat P2;
    cv::Mat Q;

    cv::Rect RL;
    cv::Rect RR;

    cv::stereoRectify( K0, kk0, K1, kk1, imgsize, R, T, R1, R2, P1, P2, Q, 0, 1.0, imgsize,  &RL, &RR );

    std::cout << "Results with OpenCV " << CV_VERSION_MAJOR << "." << CV_VERSION_MINOR << "." << CV_VERSION_REVISION << std::endl;
    std::cout << "R1 = " << R1 << std::endl;
    std::cout << "R2 = " << R1 << std::endl;
    std::cout << "P1 = " << R1 << std::endl;
    std::cout << "P2 = " << R1 << std::endl;
    std::cout << " Q = " << Q << std::endl;
    std::cout << "RL = " << RL << std::endl;
    std::cout << "RR = " << RR << std::endl;

    return 0;
}

 ```

Here is the (correct) output if the above snippet is compiled against 3.4.0:

 ```
Results with OpenCV 3.4.0
R1 = [0.9989521880366085, 0.04212807677012295, -0.01788158730434815;
 -0.04239762600809163, 0.9989886214454816, -0.01497249248903642;
 0.01723273993740956, 0.01571494098324542, 0.9997279996599788]
R2 = [0.9989521880366085, 0.04212807677012295, -0.01788158730434815;
 -0.04239762600809163, 0.9989886214454816, -0.01497249248903642;
 0.01723273993740956, 0.01571494098324542, 0.9997279996599788]
P1 = [0.9989521880366085, 0.04212807677012295, -0.01788158730434815;
 -0.04239762600809163, 0.9989886214454816, -0.01497249248903642;
 0.01723273993740956, 0.01571494098324542, 0.9997279996599788]
P2 = [0.9989521880366085, 0.04212807677012295, -0.01788158730434815;
 -0.04239762600809163, 0.9989886214454816, -0.01497249248903642;
 0.01723273993740956, 0.01571494098324542, 0.9997279996599788]
 Q = [1, 0, 0, -1256.409484863281;
 0, 1, 0, -1009.329612731934;
 0, 0, 0, 1365.958853945157;
 0, 0, -1, 110.4263305664062]
RL = [2217 x 1830 from (135, 71)]
RR = [2275 x 1883 from (73, 128)]
 ```

And the (wrong) output with 3.4.1:

 ```
Results with OpenCV 3.4.1
R1 = [-0.9989521880366083, -0.04212807677012573, 0.01788158730434768;
 0.04239645983065649, -0.9989896825345417, 0.01490484429181281;
 0.01723560879981172, 0.01564734281550852, 0.9997290105084047]
R2 = [-0.9989521880366083, -0.04212807677012573, 0.01788158730434768;
 0.04239645983065649, -0.9989896825345417, 0.01490484429181281;
 0.01723560879981172, 0.01564734281550852, 0.9997290105084047]
P1 = [-0.9989521880366083, -0.04212807677012573, 0.01788158730434768;
 0.04239645983065649, -0.9989896825345417, 0.01490484429181281;
 0.01723560879981172, 0.01564734281550852, 0.9997290105084047]
P2 = [-0.9989521880366083, -0.04212807677012573, 0.01788158730434768;
 0.04239645983065649, -0.9989896825345417, 0.01490484429181281;
 0.01723560879981172, 0.01564734281550852, 0.9997290105084047]
 Q = [1, 0, 0, -1199.557830810547;
 0, 1, 0, -1048.840476989746;
 0, 0, 0, 1365.921907767475;
 0, 0, 0.9999999999999998, 110.5510559082031]
RL = [0 x 0 from (0, 0)]
RR = [0 x 0 from (0, 0)]
 ```

**Comment by stiv-yakovenko** [comment]
I think this is related issue: http://answers.opencv.org/question/187581/stereorectify-doesnt-work-sometimes/ Please comment if possible.

**Comment by stiv-yakovenko** [comment]
@alalek  I have built my sample code, described in the link above, with OpenCV 3.4.1 (released this 26 days ago), I still get most of the stereopairs broken.

**Comment by stiv-yakovenko** [comment]
Also checked with latest repository build of opencv -- same result.

**Comment by fbergama** [comment]
Just tried with the latest repository build of master branch. Same problem.
Don't know if it might be useful, but the error occurs only if alpha is different than 0. 

Also, it is probably related to #10791 but in my case the camera  has no distortion.

**Comment by stiv-yakovenko** [comment]
@fbergama any idea where to take working code?

**Comment by fbergama** [comment]
@stiv-yakovenko Sorry, what do you mean?

**Comment by stiv-yakovenko** [comment]
@fbergama  do you know any workaround for this rectification bug?

**Comment by fbergama** [comment]
Haven’t had time to investigate. For now I’ve just downgraded to 3.4.0

**Comment by stiv-yakovenko** [comment]
So strange, I observe same problems on both 3.4.0 and 3.4.1

**Comment by fbergama** [comment]
I guess it’s due to the rational distortion model. In my case images are not distorted

**Comment by stiv-yakovenko** [comment]
What do you mean by rational distortion?

**Comment by fbergama** [comment]
I mean the rational model implemented in OpenCV when calibrated with the flag CALIB_RATIONAL_MODEL.

For details see:
https://www.robots.ox.ac.uk/~vgg/publications/2005/Claus05a/claus05a.pdf

**Comment by stiv-yakovenko** [comment]
I've solved my problem by using stereoRectifyUndistorted instead of stereoRectify. I've created 100 random points, projected them to both images from stereopair and calculated F and also rectifying homographies. All stereo pairs look perfect with this approach. So stereoRectify definitely has a bug...

**Comment by jeremywrightkim** [comment]
I can't find stereoRectifyUndistorted function..did you mean stereoRectifyUncalibrated ?

**Comment by tpet** [comment]
I had many similar problems with stereoRectify (negativ fx, fy in P1 and P2, zero ROI's, incorrect field of view for various alphas) in opencv-python-3.4.0.12 and it all seems to be fixed after upgrading opencv-python-3.4.3.18.

**Comment by zhu-ty** [comment]
I am now facing this problem in 4.0.0, how to fix it?

**Comment by zhu-ty** [comment]
problem solved, I re-implement the C version opencv source code

**Comment by Li-Shang** [comment]
@zhu-ty Same problem in 4.1.0. Could you give me some advice on fixing it? Thanks.

**Comment by zhu-ty** [comment]
> @zhu-ty Same problem in 4.1.0. Could you give me some advice on fixing it? Thanks.

Hi, I just copy the 3.4.0 source code to use the old version ...

check here:
[StereoRectify.cpp](https://github.com/zhu-ty/StereoDepth/blob/master/Cpp/ImageRectify/StereoRectify.cpp)

**Comment by yumingww** [comment]
Hi @zhu-ty and @Li-Shang  I'm using opencv4.1.0 right now and encountering the same question when I applying a monochrome image into the function.

**Comment by hellochenwang** [comment]
With the same camera intrinsic matrix and distortion variables, stereoRectify() in 3.4.4 and 3.4.7 gives different results, and none of them appears to be correct:

In Opencv 3.4.4

R1:[0.999896853671721, 0.002053753407717491, -0.01421492575897811;
 -0.001998376861133589, 0.999990363955317, 0.003908770450311316;
 0.01422281643395065, -0.003879960496271806, 0.9998913227942493]
R2:[0.9999779479868923, 0.00254726430537831, -0.006133105614847437;
 -0.002571273117301952, 0.9999890504487179, -0.003909928033559669;
 0.006123078839976506, 0.003925611701368454, 0.999973548389301]
P1:[908.2936897395464, 0, 326.1423835754395, 0;
 0, 908.2936897395464, 249.1884441375732, 0;
 0, 0, 1, 0]
P2:[908.2936897395464, 0, 326.1423835754395, -54484.83516184997;
 0, 908.2936897395464, 249.1884441375732, 0;
 0, 0, 1, 0]
Q:[1, 0, 0, -326.1423835754395;
 0, 1, 0, -249.1884441375732;
 0, 0, 0, 908.2936897395464;
 0, 0, 0.01667057791477966, 0]

In Opencv 3.4.7

R1:[0.999896853671721, 0.002053753407717509, -0.01421492575897819;
 -0.001998210979511085, 0.9999903186351127, 0.003920432237275809;
 0.01422283974016211, -0.003891623438360741, 0.9998912771380395]
R2:[0.9999779479868922, 0.002547264305378328, -0.006133105614847516;
 -0.00257120170332341, 0.9999890961653295, -0.003898265287061783;
 0.006123108828459249, 0.003913948774068135, 0.999973593922994]
P1:[811.2993931021877, 0, 326.1423835754395, 0;
 0, 811.2993931021877, 249.1768836975098, 0;
 0, 0, 1, 0]
P2:[811.2993931021877, 0, 326.1423835754395, -48666.54277071658;
 0, 811.2993931021877, 249.1768836975098, 0;
 0, 0, 1, 0]
Q:[1, 0, 0, -326.1423835754395;
 0, 1, 0, -249.1768836975098;
 0, 0, 0, 811.2993931021877;
 0, 0, 0.01667057791477966, -0]

The issue i encountered is that the stereo images don't get rectified correctly:
![image](https://user-images.githubusercontent.com/16633061/64074014-2eb31000-ccd8-11e9-9db4-3711f0a3c5e3.png)

Where in Matlab, with the same camera intrinsic matrices and distortions, R and T, the stereo  images get rectified correctly.

**Comment by stiv-yakovenko** [comment]
@hellochenwang, may be you can find exact version of opencv where the algorithm got spoilt?

**Comment by hellochenwang** [comment]
@stiv-yakovenko  i tried 2.4, 3.4.0 and 3.4.7, none of them rectified the images. if you look at the data, R1, R2 are the same across these versions. P1, P2 and Q are all slightly different among these versions. i.e. f, Cx1, Cy1, Cx2, Cy2 are calculated differently among these versions.

I also tried the stereoRectifyUndistorted way, getting R1, R2, P1, P2 through H1 and H2, it also only undistorted the images(and looked different than stereoRectify's result), rectification still didn't happen. how did you generate the 100 random points? I used the same points from chessboard corners.

`
3.4.7
R1:[0.9999776678389756, 0.003866499441780588, -0.005451055438198469;
 -0.003875921199231103, 0.9999910112256485, -0.001718924885548591;
 0.00544436021778072, 0.001740014359572358, 0.9999836655125161]
R2:[0.9999917217889253, 0.004059096422512854, 0.000283001507847639;
 -0.004059579798160889, 0.9999902652963838, 0.001728908421624442;
 -0.0002759809469227705, -0.001730042976559704, 0.9999984653917307]
P1:[909.4462728713727, 0, 317.9187049865723, 0;
 0, 909.4462728713727, 243.5866050720215, 0;
 0, 0, 1, 0]
P2:[909.4462728713727, 0, 317.9187049865723, -54.57683975635572;
 0, 909.4462728713727, 243.5866050720215, 0;
 0, 0, 1, 0]
Q:[1, 0, 0, -317.9187049865723;
 0, 1, 0, -243.5866050720215;
 0, 0, 0, 909.4462728713727;
 0, 0, 16.66359351203481, -0]

3.4.0

R1:[0.9999776678389756, 0.003866499441780588, -0.005451055438198469;
 -0.003875921199231103, 0.9999910112256485, -0.001718924885548591;
 0.00544436021778072, 0.001740014359572358, 0.9999836655125161]
R2:[0.9999917217889253, 0.004059096422512854, 0.000283001507847639;
 -0.004059579798160889, 0.9999902652963838, 0.001728908421624442;
 -0.0002759809469227705, -0.001730042976559704, 0.9999984653917307]
P1:[909.1166950086803, 0, 317.7264938354492, 0;
 0, 909.1166950086803, 243.917818069458, 0;
 0, 0, 1, 0]
P2:[909.1166950086803, 0, 317.7264938354492, -54.55706143768428;
 0, 909.1166950086803, 243.917818069458, 0;
 0, 0, 1, 0]
Q:[1, 0, 0, -317.7264938354492;
 0, 1, 0, -243.917818069458;
 0, 0, 0, 909.1166950086803;
 0, 0, 16.66359351203481, -0]


 2.4.0

 R1:[0.9999776678389756, 0.003866499441780588, -0.005451055438198469;
 -0.003875921199231103, 0.9999910112256485, -0.001718924885548591;
 0.00544436021778072, 0.001740014359572358, 0.9999836655125161]
R2:[0.9999917217889253, 0.004059096422512854, 0.000283001507847639;
 -0.004059579798160889, 0.9999902652963838, 0.001728908421624442;
 -0.0002759809469227705, -0.001730042976559704, 0.9999984653917307]
P1:[909.3446133446074, 0, 317.9754257202148, 0;
 0, 909.3446133446074, 243.6130523681641, 0;
 0, 0, 1, 0]
P2:[909.3446133446074, 0, 317.9754257202148, -54.57073905984679;
 0, 909.3446133446074, 243.6130523681641, 0;
 0, 0, 1, 0]
Q:[1, 0, 0, -317.9754257202148;
 0, 1, 0, -243.6130523681641;
 0, 0, 0, 909.3446133446074;
 0, 0, 16.66359351203481, -0]`
